pkg/controllers/management/cis/scheduled_scan.go

```diff
@@ -0,0 +1,279 @@
+	managementClient  managementv3.Interface
+	clusterClient     v3.ClusterInterface
+	clusterLister     v3.ClusterLister
+	clusterScanClient v3.ClusterScanInterface
+	clusterScanLister v3.ClusterScanLister
+	clustersMap       map[string]*scheduleInfo
+}
+
+func newScheduleScanHandler(
+	managementClient managementv3.Interface,
+	clusterClient v3.ClusterInterface,
+	clusterLister v3.ClusterLister,
+	clusterScanClient v3.ClusterScanInterface,
+	clusterScanLister v3.ClusterScanLister,
+) *scheduledScanHandler {
+	return &scheduledScanHandler{
+		managementClient:  managementClient,
+		clusterClient:     clusterClient,
+		clusterLister:     clusterLister,
+		clusterScanClient: clusterScanClient,
+		clusterScanLister: clusterScanLister,
+		clustersMap:       make(map[string]*scheduleInfo),
+	}
+}
+
+func (i *scheduleInfo) String() string {
+	return fmt.Sprintf("cronSchedule: %v lastAt: %v nextAt: %v",
+		i.cronSchedule, i.lastScanLaunchedAt.String(), i.nextScanAt.String())
+}
+
+func (ssh *scheduledScanHandler) clusterSync(key string, cluster *v3.Cluster) (runtime.Object, error) {
+	if cluster == nil ||
+		cluster.DeletionTimestamp != nil ||
+		cluster.Spec.ScheduledClusterScan == nil {
+		return nil, nil
+	}
+
+	ssh.Lock()
+	defer ssh.Unlock()
+	clusterInfo, ok := ssh.clustersMap[cluster.Name]
+	if cluster.Spec.ScheduledClusterScan.Enabled {
+		if !ok {
+			// Do stuff to enable scheduling
+			cronSchedule := getCronSchedule(cluster)
+			clusterInfo = &scheduleInfo{
+				cronSchedule: cronSchedule,
+			}
+			logrus.Infof("scheduledScanHandler: clusterSync: Scheduled Scan enabled for cluster %v: %v",
+				cluster.Name, cronSchedule)
+			if err := scheduleScan(cluster, ssh.clusterScanClient, clusterInfo); err != nil {
+				return nil, fmt.Errorf("scheduledScanHandler: clusterSync: error scheduling scan: %v", err)
+			}
+			ssh.clustersMap[cluster.Name] = clusterInfo
+		} else {
+			if cluster.Spec.ScheduledClusterScan.ScheduleConfig != nil &&
+				cluster.Spec.ScheduledClusterScan.ScheduleConfig.CronSchedule != "" &&
+				cluster.Spec.ScheduledClusterScan.ScheduleConfig.CronSchedule != clusterInfo.cronSchedule {
+				clusterInfo.cronSchedule = cluster.Spec.ScheduledClusterScan.ScheduleConfig.CronSchedule
+				logrus.Infof("scheduledScanHandler: clusterSync: cron schedule modified for cluster %v: %v",
+					cluster.Name, clusterInfo.cronSchedule)
+			}
+		}
+	} else {
+		if ok {
+			logrus.Infof("scheduledScanHandler: clusterSync: Scheduled Scan disabled for cluster: %v",
+				cluster.Name)
+			delete(ssh.clustersMap, cluster.Name)
+		}
+	}
+	return nil, nil
+}
+
+func (ssh *scheduledScanHandler) sync(key string, _ *v3.ClusterScan) (runtime.Object, error) {
+	if !strings.HasPrefix(key, ScheduledScanKeyPrefix) {
+		return nil, nil
+	}
+	splits := strings.Split(strings.TrimPrefix(key, ScheduledScanKeyPrefix), separator)
+	clusterID := splits[0]
+	pastCronSchedule := splits[1]
+	ssh.Lock()
+	defer ssh.Unlock()
+	clusterInfo, ok := ssh.clustersMap[clusterID]
```

Comment:
We should release the lock here after reading from the map. This method cleans up old scans which can potentially end up sleeping for retries and <defer ssh.Unlock()> wont be called until that returns. So better to unlock here.

Reply:
Also keep the defer

Reply:
I will change to explicitly unlock, I don't think we can do both unlock and defer unlock